### PR TITLE
[1617] Add ConstnativeHistogram

### PR DESCRIPTION
**Pull request by shivanthzen**
Fixes: #1617 
Add exported ConstNativeHistogram - this can be used to build native histograms if one already has the relevant data without calling the observe function. A viable use case is while converting exponential histograms to nativehistograms.

### CODE REVIEW

**Comment by shivanthzen** [comment]
@ArthurSens Maybe you can take a look at this PR.

File: prometheus/histogram.go
Comment (ArthurSens) [inline prometheus/histogram.go:1857]:
```suggestion
func NewConstNativeHistogram(desc *Desc, count uint64, sum float64, postiveBuckets, negativeBuckets map[int]int64, zeroBucket uint64,
```

Comment (ArthurSens) [reply prometheus/histogram.go:1857]: Could you also add.MustNewConstNativeHistogram that re uses the one you just added and panics if there's an error?

This would be consistent with all the other NewCost... functions that we have :)

File: prometheus/histogram_test.go
Comment (ArthurSens) [inline prometheus/histogram_test.go:1468]: I'm struggling a bit to understand what is being tested here 🤔 

A new constant metric is a fairly simple operation and constant metrics do not have any methods that change their internal fields such as `Observe()`, `Add()` or `Inc()`.  It's so simple that we don't even have tests for `NewConstHistogram` 🤔 

I'm not saying tests aren't useful here, I think it's worth verifying valid labels. I'm not a native histogram expert but maybe there are other things we need to make sure are correct?

File: prometheus/histogram.go
Comment (ArthurSens) [inline prometheus/histogram.go:1848]:
```suggestion
```

They are constant and we can't add extra observations, which means we don't need to care about bucket increase or resets

File: prometheus/histogram.go
Comment (ArthurSens) [inline prometheus/histogram.go:1849]:
```suggestion
	createdTimestamp                       time.Time
```

File: prometheus/histogram.go
Comment (ArthurSens) [inline prometheus/histogram.go:1861]:
```suggestion
	createdTimestamp time.Time,
```

or 

```suggestion
	ct time.Time,
```

File: prometheus/histogram.go
Comment (ArthurSens) [inline prometheus/histogram.go:1863]:
```suggestion
) (constNativeHistogram, error) {
```

I believe we need to do some validation before returning the object. The first thing I can remember is validating labels, but maybe there are more stuff we need to validate?

**Review by ArthurSens — COMMENTED** [review]
Thanks for working on this! I've added a few comments but I'm not a native histogram expert, it would be nice if @beorn7 or @krajorama could also take a look here.

question: Have you tried using your commit on OTel's side to make sure you can translate their format into ours?

**Comment by beorn7** [comment]
Will have a look ASAP (which won't be before next week, sorry).

File: prometheus/histogram_test.go
Comment (shivanthzen) [reply prometheus/histogram_test.go:1468]: THis test is an exact copy of existing test cases for native histograms in the same file.  https://github.com/prometheus/client_golang/blob/1eea8641d00e125557cf50c8123840608c1cba89/prometheus/histogram_test.go#L471 .  The tests ensure that native histograms serialised from an `observe()`ed nativehistogram is the same as the one created from constNativeHistograms - Or that serialisation of constNativehistogram is as it exists today .  Agree that this isn't really testing all the scenarios, I'll add more tests.

File: prometheus/histogram.go
Comment (krajorama) [inline prometheus/histogram.go:1855]: Needs doc string to explain the arguments and the fact that this is exposed as a counter type histogram, treated as a counter on scrape.

**Review by krajorama — COMMENTED** [review]
Hi, I understand the use case of wanting to expose OTEL exponential histograms from the Prometheus exporter in the collector. How would you use it actually? Constantly register and unregister the metric to update the exposed values? Or just use it's `Write()` method to get to the data?

In any case, the naming is going to be weird as you're not actually using it as a const :)

If I understand correctly, currently it will only work for counter histograms, not gauge histograms due to limitations in client_golang.

I'm a little afraid of abuse of this interface, but don't see true stoppers for now, will consult @beorn7 .

**Review by beorn7 — CHANGES_REQUESTED** [review]
Thanks @shivanthzen for tackling this, and thanks for the review so far @krajorama and @ArthurSens . If I haven't commented on a previous review comment, I agree with it. :)

File: prometheus/histogram.go
Comment (krajorama) [reply prometheus/histogram.go:1965]: ack

**Comment by shivanthzen** [comment]
Thanks for the comments. I'll resolve them this weekend

File: prometheus/histogram.go
Comment (shivanthzen) [reply prometheus/histogram.go]: Tried this.  Wouldn't this blow up the argument list? Is it better to create a seperate a constNaticveGuageHistogram ?

File: prometheus/histogram.go
Comment (shivanthzen) [reply prometheus/histogram.go]: Just noticed [this](https://github.com/prometheus/client_golang/blob/02b8952e721c7dc7725c4ecba742f74e980f8270/prometheus/histogram_test.go#L678) test case. Will the validation -`Count should match the sum of bucket counts` be valid in this case as well ? It seems Nan observation increments the total count, but doesn't fall into any of the buckets.

File: prometheus/histogram.go
Comment (ArthurSens) [inline prometheus/histogram.go:1880]:
```suggestion
		return nil, fmt.Errorf("invalid native histogram schema")
```

File: prometheus/histogram.go
Comment (ArthurSens) [inline prometheus/histogram.go:1884]: Do we need to validate Sum as well?

File: prometheus/histogram.go
Comment (ArthurSens) [inline prometheus/histogram.go:1927]: hey, could you clarify what `Any` means here?

**Review by ArthurSens — COMMENTED** [review]
I had just a quick look today :)

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1860]: Needs doc comment.

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1855]: The count of observations can actually be greater than the sum of all bucket counts because NaN observations go into no bucket (but then the sum of observations must be NaN).

So these are the conditions to test:
- If the sum of observations is NaN, the number of observations must be greater or equal to the sum of all bucket counts.
- Otherwise, the number of observations must be equal to the sum of all bucket counts (what you are validating right now).

As a side note, let's be more explicit in the error message. "bucket sums do not add up to total count" could mean many things. How about "the sum of all bucket populations exceeds the count of observations"?

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1846]: Maybe `bucketPopulationSum` or just `bucketSum`? `bucketCountSum` could also mean the total number of buckets. (I assume `bucktCountSum` is a typo. I wouldn't make a variable name harder to read by saving only one letter in an abbreviation attempt.)

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1868]: The function name makes it clear that we are dealing with native histograms. Let's use shorter names here:

```suggestion
	schema int32,
	zeroThreshold float64,
```

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1870]: As above, let's just call this `exemplars`.

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1879]: I'm wondering if we should finally create constants for the max and min schema. This PR adds yet another occurrence.

File: prometheus/histogram.go
Comment (beorn7) [reply prometheus/histogram.go:1884]: Not really. Sum could by anything. I mean, there could be consistency checks like the sum cannot be negative if there are only positive observations, or the sum has to be zero without any observations, and many more. But I don't know if that would be very helpful. We should only check for conditions that would make Prometheus reject the histogram upon ingestion.

File: prometheus/histogram.go
Comment (beorn7) [reply prometheus/histogram.go:1927]: Probably "positive or negative".
I would just call it `makeBuckets`.

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1894]: Exemplars for native histograms MUST have a timestamp. We should validate that.

File: prometheus/histogram.go
Comment (beorn7) [reply prometheus/histogram.go:1860]: We should also have a version `MustNewConstNativeHistogram` (which wraps `NewConstNativeHistogram`, panicking where `NewConstNativeHistogram` would have returned an error).

**Review by beorn7 — CHANGES_REQUESTED** [review]
Looks good in general.

Despite the frequent nit pickings about naming etc., I haven't yet checked the technical correctness very deeply yet (although it looks fine as far as I can see right now).

File: prometheus/histogram.go
Comment (ArthurSens) [reply prometheus/histogram.go:1927]: +1 for makeBuckets

Comment (shivanthzen) [reply prometheus/histogram.go:1927]: Hmm, makebuckets alreadyexists in the namespace. I could do `makeBucketfrommap`

File: prometheus/histogram.go
Comment (shivanthzen) [reply prometheus/histogram.go:1899]: Added the docstring

**Comment by ArthurSens** [comment]
Oh, looks like some commits weren't signed and there's one test failing 🤔 

The test seems to be flaky, it's also failing in other PRs

File: prometheus/histogram.go
Comment (shivanthzen) [reply prometheus/histogram.go]: I'm still unclear about including a ` gauge histogram` . As per my understanding , we will need to add  postiveBuckets, negativeBuckets of type  map[int]float64 as an argument (or use a generic parameter).

**Comment by shivanthzen** [comment]
I can sign the commits once the PR is approved, It seems  signing will need a force push

File: prometheus/histogram.go
Comment (beorn7) [reply prometheus/histogram.go]: It sounds like you are thinking about a float histogram. That's something we might want to add later (ideally together with the option of creating classic float histograms). However, the use case for float histograms in instrumentation is very rare.

I am in fact talking about a gauge histogram. It merely requires to set a different `MetricType` in the `MetricFamily`. However, I've just realized that we currently have no way of communicating the difference of counter histogram (i.e. the "normal" histogram) and gauge histogram during metrics collection. For that, we needed `Desc` to have a metric type field (which was the plan for quite some time).

So I'm pretty sure we cannot solve this problem in this PR, because it needed too much re-architecting.

But if we assume that the difference between counter histogram and gauge histogram will be in the `Desc`, we actually don't need a bool parameter here, but just have to later allow the option of creating "typed" `Desc`s.

In conclusion, let's not do gauge histograms right now.

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1908]: We shouldn't use naked `dto.Exemplar` here. Check out how it is done for `NewMetricWithExemplars`, i.e. there is a `prometheus.Exemplar` and a `newExemplar` constructor to create the `dto.Exemplar`.

Looking at `NewMetricWithExemplars`, however, I'm wondering if we should even allow an exemplar argument here. We could also implement native histogram handling in `withExemplarsMetric` (strictly speaking, we have to do that anyway), and then all `NewConst...` functions would be consistent (as in: not having exemplars, but you use the `NewMetricWithExemplars` to wrap them).

@ArthurSens @bwplotka WDYT?

File: prometheus/histogram_test.go
Comment (beorn7) [inline prometheus/histogram_test.go:1459]: Nit:
```suggestion
func syncMapToMap(syncmap *sync.Map) (m map[int]int64) {
```

**Review by beorn7 — CHANGES_REQUESTED** [review]
Thanks for addressing the feedback.

I have no realized something about exemplars we had not discussed before.

I have also realized we cannot do gauge histograms right now. Needs more discussion and planning, which is out of scope of this PR.

File: prometheus/histogram.go
Comment (shivanthzen) [reply prometheus/histogram.go:1872]: About Nil timestamp:
From https://pkg.go.dev/google.golang.org/protobuf/types/known/timestamppb#Timestamp.CheckValid , 
```
CheckValid returns an error if the timestamp is invalid. In particular, it checks whether the value represents a date that is in the range of 0001-01-01T00:00:00Z to 9999-12-31T23:59:59Z inclusive. An error is reported for a nil Timestamp.
```
More  here: https://github.com/protocolbuffers/protobuf-go/blob/158d2b331a354322bceddf905a52b129d1a740d7/types/known/timestamppb/timestamp.pb.go#L215
In short,
`e.GetTimestamp().IsValid()` will return false for a nil timestamp. No explicit check required.

File: prometheus/histogram.go
Comment (shivanthzen) [reply prometheus/histogram.go]: That's true. I was indeed confused about a float histogram and gauge histogram.

**Comment by shivanthzen** [comment]
Force pushed with signed commits.

File: prometheus/histogram.go
Comment (shivanthzen) [reply prometheus/histogram.go]: updated to `the sum of counts in the bucket do not equal count when sum is not Nan`

File: prometheus/histogram.go
Comment (shivanthzen) [reply prometheus/histogram.go:1908]: I'll wait for a consensus here.

Comment (ArthurSens) [reply prometheus/histogram.go:1908]: Sounds reasonable to me :), let's keep this pattern.

**Comment by ArthurSens** [comment]
@shivanthzen, looks like you accidentally added unrelated commits to your PR, could you keep only the ones that you've authored here?

File: prometheus/histogram.go
Comment (ArthurSens) [reply prometheus/histogram.go:1908]: To unblock this PR and make it simpler, maybe here we could leave exemplars out. We deal with exemplars in a follow-up PR that implements `.withExemplars(e []*dto.Exemplar)`. Would that work for both of you @beorn7 @shivanthzen?

Comment (beorn7) [reply prometheus/histogram.go:1908]: Yes, that's essentially what I sketched up as plan B above. As said, we have to make `NewMetricWithExemplars` work with native histograms anyway, and then we don't really need it derictly in `NewConstNativeHistogram` anymore.

**Comment by shivanthzen** [comment]
Removed other commits from the branch

**Review by krajorama — COMMENTED** [review]
couple of minor comments

**Comment by kakkoyun** [comment]
@beorn7 could you have another look at it?

File: prometheus/histogram.go
Comment (shivanthzen) [reply prometheus/histogram.go:1885]: Is the suggestion to add 

```Note that Prometheus Native Histograms buckets are indexed by upper boundary while OTEL Exponential Histograms are indexed by lower boundary, the result being that the Offset fields are different-by-one.`` ?

**Comment by beorn7** [comment]
This is very close to the top of my review queue now. Will have a look soon, hopefully today.

File: prometheus/histogram.go
Comment (beorn7) [reply prometheus/histogram.go:1885]: I think we should just add the link @krajorama suggested. Something like: "See https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/#exponential-histograms for more details about the conversion from OTel to Prometheus."

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1889]: Some wordsmithing, but also note the indentation so that Godoc will actually render this as a list:

```suggestion
//  - the length of labelValues is not consistent with the variable labels in Desc or if Desc is invalid.
//  - the schema passed is not between 8 and -4
//  - the sum of counts in all buckets including the zero bucket does not equal the count if sum is not NaN (or exceeds the count if sum is NaN)
```

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1944]: This needs a doc comment. (A simple one, see the other `MustNewConst...` functions al reference.)

**Review by beorn7 — CHANGES_REQUESTED** [review]
Just some minor things about doc comments left.

**Comment by beorn7** [comment]
Looks good to me, but there are lint warnings now. (They are coming from a recently added check, but they are still relevant.) Also, DCO signing is missing. @shivanthzen could you address the lint warnings and sign the commits? (Feel free to squash commits into fewer or just one before doing that.)

**Review by beorn7 — APPROVED** [review]
Thank you very much. I leave it to the maintainers to do the final approval and merge.

**Review by ArthurSens — APPROVED** [review]
LGTM, thank you all for the work here :)

Follow-up here would be to tackle the addition of exemplars!

**Comment by beorn7** [comment]
> Follow-up here would be to tackle the addition of exemplars!

To clarify: This means making [withExemplarsMetric.Write](https://github.com/prometheus/client_golang/blob/291b0b0c426b6fb4025d8a3c2c8e7385026ba1d8/prometheus/metric.go#L180) aware of native histogram exemplars. Then, we only need to update the doc comment of [NewMetricWithExemplar](https://github.com/prometheus/client_golang/blob/291b0b0c426b6fb4025d8a3c2c8e7385026ba1d8/prometheus/metric.go#L233) and everything should work fine.

**Pull request merged**
